From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aswin-mcw <aswin.kalies@multicorewareinc.com>
Date: Wed, 9 Sep 2026 13:14:15 +0530
Subject: [PATCH 037/109] Cygwin: _endian.h: Add AArch64 implementations
 for `ntohl` and `ntohs`

(cherry picked from commit f2083179f4e30e3f7275abe4fd0ea2d87f84d136)
---
 winsup/cygwin/include/machine/_endian.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/winsup/cygwin/include/machine/_endian.h b/winsup/cygwin/include/machine/_endian.h
index dbd4429b8..622d7a2e9 100644
--- a/winsup/cygwin/include/machine/_endian.h
+++ b/winsup/cygwin/include/machine/_endian.h
@@ -26,16 +26,26 @@ _ELIDABLE_INLINE __uint16_t __ntohs(__uint16_t);
 _ELIDABLE_INLINE __uint32_t
 __ntohl(__uint32_t _x)
 {
+#if defined(__x86_64__)
 	__asm__("bswap %0" : "=r" (_x) : "0" (_x));
+#elif defined(__aarch64__)
+	__asm__("rev %w0, %w0" : "=r" (_x) : "0" (_x));
+#endif
 	return _x;
 }
 
 _ELIDABLE_INLINE __uint16_t
 __ntohs(__uint16_t _x)
 {
+#if defined(__x86_64__)
 	__asm__("xchgb %b0,%h0"		/* swap bytes		*/
 		: "=Q" (_x)
 		:  "0" (_x));
+#elif defined(__aarch64__)
+	__asm__("uxth %w0, %w0\n\t"
+		"rev16 %w0, %w0"
+		: "+r" (_x));
+#endif
 	return _x;
 }
 
